pandas/io/stata.py

```diff
@@ -1162,21 +1170,48 @@ def __init__(
         self._lines_read = 0
 
         self._native_byteorder = _set_endianness(sys.byteorder)
-        with get_handle(
-            path_or_buf,
+
+    def _ensure_open(self) -> None:
+        """
+        Ensure the file has been opened and its header data read.
+        """
+        if not hasattr(self, "_path_or_buf"):
+            self._open_file()
+
+    def _open_file(self) -> None:
+        """
+        Open the file (with compression options, etc.), and read header information.
+        """
+        if not self._entered:
+            warnings.warn(
+                "StataReader is being used without using a context manager. "
```

Comment:
Probably missed the discussion, but why can't we just raise an exception to say "use the context manager version" instead?

Reply:
The context is in e.g. these comments https://github.com/pandas-dev/pandas/pull/48922#issuecomment-1279706702 https://github.com/pandas-dev/pandas/pull/48922#issuecomment-1279735613 on the v1 PR of this where @bashtage and @jreback are strongly against breaking (already somewhat broken, IMO) user code that misused StataReader by not using it as a CM.